sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java

```diff
@@ -270,13 +271,40 @@ private void initBatch(
         vectors[i + partitionIdx].setIsConstant();
       }
     }
+
+    // For Parquet tables whose columns have associated DEFAULT values, this reader must return
+    // those values instead of NULL when the corresponding columns are not present in storage (i.e.
+    // belong to the 'missingColumns' field in this class).
+    ColumnVector[] finalColumns = new ColumnVector[sparkSchema.fields().length];
+    for (int i = 0; i < columnVectors.length; i++) {
+      Object defaultValue = sparkRequestedSchema.existenceDefaultValues()[i];
+      if (defaultValue == null) {
+        finalColumns[i] = vectors[i];
+      } else {
+        WritableColumnVector writable;
+        if (memMode == MemoryMode.OFF_HEAP) {
+          writable = new OffHeapColumnVector(capacity, vectors[i].dataType());
+        } else {
+          writable = new OnHeapColumnVector(capacity, vectors[i].dataType());
+        }
+        Optional<Integer> appended = writable.appendObjects(capacity, defaultValue);
+        if (!appended.isPresent()) {
+          throw new IOException("Cannot assign default column value to result column batch in " +
+            "vectorized Parquet reader because the data type is not supported: " + defaultValue);
+        }
```

Comment:
If `isPresent` is false, throwing exception here, but in Orc reader it silently ignores that and continues to pull nulls. Why the difference?

Reply:
Good catch, I updated the behavior to be the same.